src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+        var unchangedSegment = { type: "unchanged"};
+        unchangedSegment[key] = leftValue;
+        segments.push(unchangedSegment);
+    }
+    else {
+        if ((!isArray && leftValue !== undefined) || (isArray && leftValue.length)) {
+            var removedSegment = { type: "removed" };
+            removedSegment[key] = leftValue;
+            segments.push(removedSegment);
+        }
+        if ((!isArray && rightValue !== undefined) || (isArray && rightValue.length)) {
+            var addedSegment = { type: "added"};
+            addedSegment[key] = rightValue;
+            segments.push(addedSegment);
+        }
+    }
+    return segments;
+};
+
+/**
+ *
+ * Find the longest matching "phrase" in two strings, which prefers one or more whole words.
+ *
+ * @param leftString {String} - One of the two strings, which will be used to search for matching patterns.
+ * @param rightString {String} - The other string, which will be compared against all possible sub-patterns.
+ * @return {Object} - See below.
+ *
+ * The object returned has three keys:
+ *
+ * `leftIndex`:  The position of the matching segment in `leftString`.  If there is no matching segment, this will be `-1`.
+ * `rightIndex`: The position of the matching segment in `rightString`.  If there is no matching segment, this will be `-1`.
+ * `segment`:    The longest matching string common to both strings.  If there are multiple matching strings of the same length, the first is returned.
+ *
+ */
+gpii.diff.longestCommonStringPhrase = function (leftString, rightString) {
+    var matchDef = {
+        leftIndex: -1,
+        rightIndex: -1,
+        segment: ""
+    };
+
+    if (leftString && rightString) {
+        var leftWordPhraseDefs = gpii.diff.extractPhraseSegments(leftString);
+
+        var firstMatchDef = fluid.find(leftWordPhraseDefs, function (phraseDef) {
+            var rightIndex = rightString.indexOf(phraseDef.value);
+            if (rightIndex !== -1) {
+                return {
+                    segment: phraseDef.value,
+                    leftIndex: phraseDef.index,
+                    rightIndex: rightIndex
+                };
+            }
+        });
+        if (firstMatchDef) {
+            matchDef = firstMatchDef;
+        }
+    }
+    return matchDef;
+};
+
+gpii.diff.sortPhraseDefsByLength = function (a, b) {
+    if (a.value.length > b.value.length) {
+        return -1;
+    }
+    else if (a.value.length === b.value.length) {
+        if (a.index < b.index) {
+            return -1;
+        }
+        else if (a.index === b.index) {
+            return 0;
+        }
+        else if (b.index < a.index) {
+            return 1;
+        }
+    }
+    else {
+        return 1;
+    }
+};
+
+/**
+ *
+ * Break down a string into an array of "phrase segments", ordered from longest (the original string) to shortest (the
+ * shortest individual word).  Each "phrase segment" looks like:
```

Comment:
The reader should be warned that this is an algorithm which is quadratic in both time and space and should not be called with terribly long input.

Reply:
It should be more reasonable now, I will confirm with the very long comparisons I need to make in the package that uses this to diff longer blocks of text.